packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1374 @@
+          addedBoundElements.find((x) => x.type === "text")
+        ) {
+          nextBoundElements = ElementsChange.unbindExistingTextElements(
+            nextBoundElements ?? [],
+            elements,
+          );
+        }
+
+        const mergedBoundElements = Object.values(
+          Delta.merge(
+            arrayToObject(nextBoundElements ?? [], (x) => x.id),
+            arrayToObject(addedBoundElements ?? [], (x) => x.id),
+            arrayToObject(removedBoundElements ?? [], (x) => x.id),
+          ),
+        );
+
+        nextBoundElements = mergedBoundElements.length
+          ? mergedBoundElements
+          : null;
+      }
+
+      let nextGroupIds = groupIds;
+      if (addedGroupIds?.length || removedGroupIds?.length) {
+        const mergedGroupIds = Object.values(
+          Delta.merge(
+            arrayToObject(groupIds ?? []),
+            arrayToObject(addedGroupIds ?? []),
+            arrayToObject(removedGroupIds ?? []),
+          ),
+        );
+        nextGroupIds = mergedGroupIds;
+      }
+
+      const mergedPartial: ElementPartial = {
+        ...directlyApplicablePartial,
+        boundElements: nextBoundElements,
+        groupIds: nextGroupIds,
+      };
+
+      if (!flags.containsVisibleDifference) {
+        // Strip away fractional as even if it would be different, it doesn't have to result in visible change
+        const { index, ...rest } = mergedPartial;
+        const containsVisibleDifference =
+          ElementsChange.checkForVisibleDifference(element, rest);
+
+        flags.containsVisibleDifference = containsVisibleDifference;
+      }
+
+      if (!flags.containsZindexDifference) {
+        flags.containsZindexDifference =
+          delta.deleted.index !== delta.inserted.index;
+      }
+
+      const updatedElement = newElementWith(element, mergedPartial);
+      elements.set(updatedElement.id, updatedElement);
+
+      return updatedElement;
+    };
+
+  private static createGetter =
+    (
+      elements: SceneElementsMap,
+      snapshot: Map<string, OrderedExcalidrawElement>,
+      flags: {
+        containsVisibleDifference: boolean;
+        containsZindexDifference: boolean;
+      },
+    ) =>
+    (id: string, partial: ElementPartial) => {
+      let element = elements.get(id);
+
+      if (!element) {
+        // Always fallback to the local snapshot, in cases when we cannot find the element in the elements array
+        element = snapshot.get(id);
+
+        if (element) {
+          // As the element was brought from the snapshot, it automatically results in possible zindex difference
+          flags.containsZindexDifference = true;
+
+          // As the element was force deleted, we need to check if adding it back results in a visible change
+          if (
+            partial.isDeleted === false ||
+            (partial.isDeleted !== true && element.isDeleted === false)
+          ) {
+            flags.containsVisibleDifference = true;
```

Comment:
Test all these cases - restoring from snapshot for added/updated/removed, setting flags for zindex differences & visible differences.

Reply:
Tested as part of the scene drag&drop